File: JavaCheatSheet/src/main/java/com/github/jameswdelancey/javaCheatSheet/Part7.java
package com.github.jameswdelancey.javaCheatSheet;

public class Part7 {
    public static void main(String[] args) {
        Hands hands = new Hands();
        hands.report();
        hands = hands.juggle();
        hands.report();
    }
}

class Hands {
    private Fruit left;
    private Fruit right;

    public Hands() {
        this.left = Fruit.Apple;
        this.right = Fruit.Banana;
    }

    public Hands juggle() {
        System.out.println("Let's juggle!");
        Fruit air = this.left;
        this.left = this.right;
        this.right = air;
        return this;
    }

    @Override
    public String toString() {
        return "Left hand is holding " + left + "\n" +
                "Right hand is holding " + right;
    }

    public void report() {
        System.out.println(this);
    }
}

enum Fruit {
    Apple,
    Banana;

    @Override
    public String toString() {
        return this == Apple ? "an apple" : "a banana";
    }
}

File: rustCheatSheet/src/part7.rs
use Model::Hands;

mod Model {
    use std::fmt::Display;
    // pub trait Displayable {
    //     fn display(&self) -> String;
    // }
    pub enum Fruit {
        Apple,
        Banana,
    }
    // impl Displayable for Fruit {
    //     // fn display(&self) -> String {
    //     //     if let Fruit::Apple = self {
    //     //         "an apple".to_owned()
    //     //     } else {
    //     //         "a banana".to_owned()
    //     //     }
    //     // }
    //     fn display(&self) -> String {
    //         match self {
    //             Fruit::Apple => "an apple".to_owned(),
    //             Fruit::Banana => "a banana".to_owned(),
    //         }
    //     }
    // }
    // pub enum Item<T> {
    //     Something(T),
    //     Nothing,
    // }
    impl Display for Fruit {
        fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
            match self {
                Fruit::Apple => f.write_str("an apple"),
                Fruit::Banana => f.write_str("a banana"),
            }
        }
    }

    pub struct Hands {
        left: Option<Fruit>,
        right: Option<Fruit>,
    }
    impl Hands {
        pub fn new() -> Hands {
            Hands {
                left: Option::Some(Fruit::Apple),
                right: Option::Some(Fruit::Banana),
            }
        }

        pub fn juggle(mut self) -> Self {
            println!("Let's juggle!");
            let air = self.left;
            self.left = self.right;
            self.right = air;
            self
        }
        pub fn report(&self) {
            reportItem(&self.left, "Left");
            reportItem(&self.right, "Right");
        }
    }

    pub fn reportItem<T: Display>(item: &Option<T>, which: &str) {
        match item {
            Option::Some(what) => {
                println!("{} hand is holding {}", which, what)
            }
            _ => println!("{} hand is not holding anything", which),
        }
    }
}

#[test]
fn main() {
    let mut hands = Hands::new();
    hands.report();
    hands = hands.juggle();
    hands.report();
}
